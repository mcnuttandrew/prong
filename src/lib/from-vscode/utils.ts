import * as Json from "jsonc-parser";
import { JSONSchemaRef, JSONSchema } from "../JSONSchemaTypes";
import { ASTNode } from "./parser";
import * as nls from "vscode-nls";
let localize = nls.loadMessageBundle();

export function isNumber(val: any): val is number {
  return typeof val === "number";
}

export function isDefined(val: any): val is object {
  return typeof val !== "undefined";
}

export function isBoolean(val: any): val is boolean {
  return typeof val === "boolean";
}

export function isString(val: any): val is string {
  return typeof val === "string";
}

export function getNodeValue(node: ASTNode): any {
  return Json.getNodeValue(node);
}

export function asSchema(
  schema: JSONSchemaRef | undefined
): JSONSchema | undefined {
  if (isBoolean(schema)) {
    return schema ? {} : { not: {} };
  }
  return schema;
}

export function contains(
  node: ASTNode,
  offset: number,
  includeRightBound = false
): boolean {
  return (
    (offset >= node.offset && offset < node.offset + node.length) ||
    (includeRightBound && offset === node.offset + node.length)
  );
}

export function equals(one: any, other: any): boolean {
  if (one === other) {
    return true;
  }
  if (
    one === null ||
    one === undefined ||
    other === null ||
    other === undefined
  ) {
    return false;
  }
  if (typeof one !== typeof other) {
    return false;
  }
  if (typeof one !== "object") {
    return false;
  }
  if (Array.isArray(one) !== Array.isArray(other)) {
    return false;
  }

  var i: number, key: string;

  if (Array.isArray(one)) {
    if (one.length !== other.length) {
      return false;
    }
    for (i = 0; i < one.length; i++) {
      if (!equals(one[i], other[i])) {
        return false;
      }
    }
  } else {
    var oneKeys: string[] = [];

    for (key in one) {
      oneKeys.push(key);
    }
    oneKeys.sort();
    var otherKeys: string[] = [];
    for (key in other) {
      otherKeys.push(key);
    }
    otherKeys.sort();
    if (!equals(oneKeys, otherKeys)) {
      return false;
    }
    for (i = 0; i < oneKeys.length; i++) {
      if (!equals(one[oneKeys[i]], other[oneKeys[i]])) {
        return false;
      }
    }
  }
  return true;
}

export enum ErrorCode {
  Undefined = 0,
  EnumValueMismatch = 1,
  Deprecated = 2,
  UnexpectedEndOfComment = 0x101,
  UnexpectedEndOfString = 0x102,
  UnexpectedEndOfNumber = 0x103,
  InvalidUnicode = 0x104,
  InvalidEscapeCharacter = 0x105,
  InvalidCharacter = 0x106,
  PropertyExpected = 0x201,
  CommaExpected = 0x202,
  ColonExpected = 0x203,
  ValueExpected = 0x204,
  CommaOrCloseBacketExpected = 0x205,
  CommaOrCloseBraceExpected = 0x206,
  TrailingComma = 0x207,
  DuplicateKey = 0x208,
  CommentNotPermitted = 0x209,
  SchemaResolveError = 0x300,
}
