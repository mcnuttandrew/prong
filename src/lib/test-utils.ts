import { SyntaxNode } from "@lezer/common";
import { parser } from "@lezer/json";
// could also write one that is find by text content?
export function findNodeByLocation(
  text: string,
  from: number,
  to: number
): SyntaxNode | null {
  let foundNode: SyntaxNode | null = null;
  parser.parse(text).iterate({
    enter: (node) => {
      // console.log(node.from, node.to, node.node.type);
      if (node.from === from && node.to === to) {
        foundNode = node.node as unknown as SyntaxNode;
      }
    },
  });
  return foundNode;
}

export function findNodeByText(
  text: string,
  matchText: string
): SyntaxNode | null {
  let foundNode: SyntaxNode | null = null;
  parser.parse(text).iterate({
    enter: (node) => {
      // console.log(node.from, node.to, node.node.type);
      const testText = text.slice(node.from, node.to);
      if (testText === matchText) {
        foundNode = node.node as unknown as SyntaxNode;
      }
    },
  });
  return foundNode;
}
