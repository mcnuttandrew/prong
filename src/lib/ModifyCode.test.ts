import { insertSwap } from "./utils";
import { modifyCodeByCommand, MenuEvent } from "./modify-json";
import { findNodeByText } from "./test-utils";

const exampleData = `{
    "a": {
      "b": [1, 2, 
        3],
      "c": true,
    },
    "d": null,
    "e": [{ "f": 4, "g": 5 }],
    "I": "example",
  }`;

test.only("modifyCodeByCommand - decreaseItemIdx", () => {
  ["1", "2", "3", '"a"', '"d"', '"e"', '"I"'].forEach((key) => {
    const keyNode = findNodeByText(exampleData, key)!;
    const cmd = modifyCodeByCommand(
      { payload: key, type: "decreaseItemIdx" },
      keyNode,
      exampleData
    )!;
    expect(insertSwap(exampleData, cmd)).toMatchSnapshot();
  });
});

test("modifyCodeByCommand - increaseItemIdx", () => {
  ["1", "2", "3", '"a"', '"d"', '"e"', '"I"'].forEach((key) => {
    const keyNode = findNodeByText(exampleData, key)!;
    const cmd = modifyCodeByCommand(
      { payload: key, type: "increaseItemIdx" },
      keyNode,
      exampleData
    )!;
    expect(insertSwap(exampleData, cmd)).toMatchSnapshot();
  });
});

test("modifyCodeByCommand - simpleSwap", () => {
  const trueKey = findNodeByText(exampleData, "true")!;
  [
    "false",
    '"hot dog sports"',
    "0",
    '{"squid": ["dough", false, "bag", 0]}',
  ].forEach((payload) => {
    const cmd = modifyCodeByCommand(
      { payload, type: "simpleSwap" },
      trueKey,
      exampleData
    )!;
    expect(insertSwap(exampleData, cmd)).toMatchSnapshot();
  });
  // may also want to check that it swaps in arrays fine?
  // const threeKey =
});

test("modifyCodeByCommand - addElementAsSiblingInArray", () => {
  const makeCmd = (): MenuEvent =>
    copy({ type: "addElementAsSiblingInArray", payload: "-9" });
  // insert from left
  const key1 = findNodeByText(exampleData, "1")!.prevSibling!;
  const cmd1 = modifyCodeByCommand(makeCmd(), key1, exampleData)!;
  const result1 = insertSwap(exampleData, cmd1);
  expect(result1).toMatchSnapshot();

  // insert to right with trailing comma
  const exampleData2 = exampleData.replace("3", "3,");
  const key2 = findNodeByText(exampleData2, "3")!;
  const cmd2 = modifyCodeByCommand(makeCmd(), key2, exampleData)!;
  const result2 = insertSwap(exampleData, cmd2);
  expect(result2).toMatchSnapshot();

  // test empty array
  const exampleData3 = exampleData.replace("3", "[]");
  const key3 = findNodeByText(exampleData3, "[]")!.firstChild!;
  const cmd3 = modifyCodeByCommand(makeCmd(), key3, exampleData)!;
  const result3 = insertSwap(exampleData3, cmd3);
  expect(result3).toMatchSnapshot();

  // insert at each place across array
  [1, 2, 3].forEach((key) => {
    const foundKey = findNodeByText(exampleData, `${key}`)!;
    const cmd = modifyCodeByCommand(makeCmd(), foundKey, exampleData)!;
    const result = insertSwap(exampleData, cmd);
    expect(result).toMatchSnapshot();
  });

  expect(true).toBe(true);
});

test("modifyCodeByCommand - addObjectKey", () => {
  [exampleData, `{ "f": 4, "g": 5 }`].forEach((targKey) => {
    const rootKey = findNodeByText(exampleData, targKey)!;
    const cmd = modifyCodeByCommand(
      { payload: { key: `"J"`, value: "[6, 7, 8]" }, type: "addObjectKey" },
      rootKey,
      exampleData
    )!;
    expect(insertSwap(exampleData, cmd)).toMatchSnapshot();
  });

  const cmd2 = modifyCodeByCommand(
    { payload: { key: `"J"`, value: "{}" }, type: "addObjectKey" },
    findNodeByText(exampleData, exampleData)!,
    exampleData
  )!;
  const update1 = insertSwap(exampleData, cmd2);
  const newObjKey = findNodeByText(update1, "{}")!;
  const cmd3 = modifyCodeByCommand(
    { payload: { key: `"X"`, value: '"Y"' }, type: "addObjectKey" },
    newObjKey,
    update1
  )!;
  expect(insertSwap(update1, cmd3)).toMatchSnapshot();
});

const copy = (x: any) => JSON.parse(JSON.stringify(x));

test("modifyCodeByCommand - removeObjectKey", () => {
  const makeCmd = (): MenuEvent =>
    copy({ payload: "", type: "removeObjectKey" });

  // test a bunch of object keys
  ["a", "d", "I", "f", "g"].forEach((text) => {
    const key = findNodeByText(exampleData, `"${text}"`)!;
    const cmd = modifyCodeByCommand(makeCmd(), key, exampleData)!;
    expect(insertSwap(exampleData, cmd)).toMatchSnapshot();
  });

  // remove all keys from an object
  [
    ["f", "g"],
    ["g", "f"],
  ].forEach(([firstKey, secondKey]) => {
    const key1 = findNodeByText(exampleData, `"${firstKey}"`)!;
    const updated = insertSwap(
      exampleData,
      modifyCodeByCommand(makeCmd(), key1, exampleData)!
    );
    const key2 = findNodeByText(updated, `"${secondKey}"`)!;
    const cmd = modifyCodeByCommand(makeCmd(), key2, updated)!;
    expect(insertSwap(updated, cmd)).toMatchSnapshot();
  });
});

test("modifyCodeByCommand - removeElementFromArray", () => {
  const makeCmd = (): MenuEvent =>
    copy({ payload: "", type: "removeElementFromArray" });

  // test a bunch of object keys
  ["1", "2", "3", `{ "f": 4, "g": 5 }`].forEach((text) => {
    const key = findNodeByText(exampleData, `${text}`)!;
    const cmd = modifyCodeByCommand(makeCmd(), key, exampleData)!;
    expect(insertSwap(exampleData, cmd)).toMatchSnapshot();
  });

  // remove all elements in an array
  [
    ["1", "2", "3"],
    ["3", "1", "2"],
    ["3", "2", "1"],
  ].forEach(([firstKey, secondKey, thirdKey]) => {
    const key1 = findNodeByText(exampleData, `${firstKey}`)!;
    const cmd0 = modifyCodeByCommand(makeCmd(), key1, exampleData);
    const update1 = insertSwap(exampleData, cmd0!);
    // console.log(key1);
    const key2 = findNodeByText(update1, `${secondKey}`)!;
    const cmd = modifyCodeByCommand(makeCmd(), key2, update1)!;
    const update2 = insertSwap(update1, cmd);

    const key3 = findNodeByText(update2, `${thirdKey}`)!;
    const cmd2 = modifyCodeByCommand(makeCmd(), key3, update2)!;
    const update3 = insertSwap(update2, cmd2);
    expect(update3).toMatchSnapshot();
  });
});
