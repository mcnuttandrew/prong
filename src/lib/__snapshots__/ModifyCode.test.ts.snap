// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modifyCodeByCommand - addElementAsSiblingInArray (empty array) 1`] = `"{\\"example\\": [-9]}"`;

exports[`modifyCodeByCommand - addElementAsSiblingInArray 1`] = `
"{
    \\"a\\": {
      \\"b\\": [-9, 1, 2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - addElementAsSiblingInArray 2`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3, -9],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - addElementAsSiblingInArray 3`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        [-9]],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - addElementAsSiblingInArray 4`] = `
"{
    \\"a\\": {
      \\"b\\": [1, -9, 2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - addElementAsSiblingInArray 5`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, -9, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - addElementAsSiblingInArray 6`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3, -9],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - addObjectKey (boundary effects) 1`] = `
"{
    \\"test\\": null,
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - addObjectKey (buggy mark) 1`] = `"{\\"mark\\": {\\"type\\":\\"bar\\",  \\"color\\": \\"\\"  }} "`;

exports[`modifyCodeByCommand - addObjectKey (cursorPos fruit) 1`] = `
"{
  \\"fruits\\": [ \\"apple\\", \\"orange\\", \\"#c71585\\" ],
  \\"vegetables\\": [
    {
      \\"veggieStarRating\\": 0,
      \\"veggieName\\": \\"potato\\",
      \\"veggieLike\\": true
    },
    {
      \\"veggieName\\": \\"broccoli\\",
      \\"veggieLike\\": false
    }
  ]
}"
`;

exports[`modifyCodeByCommand - addObjectKey (cursorPos version) 1`] = `"{\\"arc\\": {}}"`;

exports[`modifyCodeByCommand - addObjectKey (cursorPos version) 2`] = `"{ \\"bar\\": { },     \\"point\\": { }      }"`;

exports[`modifyCodeByCommand - addObjectKey (cursorPos version) 3`] = `"{\\"bar\\": { },  \\"point\\": { }   }"`;

exports[`modifyCodeByCommand - addObjectKey (fruit insert 0) 1`] = `
"
{
  \\"fruits\\": [ \\"apple\\", \\"orange\\", \\"#c71585\\" ],
  \\"vegetables\\": [
    {
      \\"veggieStar\\": [1, 2, 3],
      \\"veggieName\\": \\"potato\\",
      \\"veggieLike\\": true 
    },
    {
      \\"veggieName\\": \\"broccoli\\",
      \\"veggieLike\\": false
    }
  ]
}"
`;

exports[`modifyCodeByCommand - addObjectKey (fruit insert 1) 1`] = `
"
{
  \\"fruits\\": [ \\"apple\\", \\"orange\\", \\"#c71585\\" ],
  \\"vegetables\\": [
    {
      \\"veggieName\\": \\"potato\\",
      \\"veggieStar\\": [1, 2, 3],
      \\"veggieLike\\": true 
    },
    {
      \\"veggieName\\": \\"broccoli\\",
      \\"veggieLike\\": false
    }
  ]
}"
`;

exports[`modifyCodeByCommand - addObjectKey (fruit insert 2) 1`] = `
"
{
  \\"fruits\\": [ \\"apple\\", \\"orange\\", \\"#c71585\\" ],
  \\"vegetables\\": [
    {
      \\"veggieName\\": \\"potato\\",
      \\"veggieLike\\": true,
          \\"veggieStar\\": [1, 2, 3]
    },
    {
      \\"veggieName\\": \\"broccoli\\",
      \\"veggieLike\\": false
    }
  ]
}"
`;

exports[`modifyCodeByCommand - addObjectKey 1`] = `
"{
    \\"J\\": [6, 7, 8],
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - addObjectKey 2`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5, \\"J\\": [6, 7, 8] }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - addObjectKey 3`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  \\"J\\": {\\"X\\": \\"Y\\"}
  }"
`;

exports[`modifyCodeByCommand - addObjectKeyEvent (legend insert) 1`] = `
"
{
  \\"padding\\": 5,
    \\"legends\\": [],
  \\"signals\\": []
}
"
`;

exports[`modifyCodeByCommand - decreaseItemIdx 1`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - decreaseItemIdx 2`] = `
"{
    \\"a\\": {
      \\"b\\": [2, 1, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - decreaseItemIdx 3`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 3, 
        2],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - decreaseItemIdx 4`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - decreaseItemIdx 5`] = `
"{
    \\"d\\": null,
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - decreaseItemIdx 6`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"d\\": null,
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - decreaseItemIdx 7`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"I\\": \\"example\\",
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
  }"
`;

exports[`modifyCodeByCommand - increaseItemIdx 1`] = `
"{
    \\"a\\": {
      \\"b\\": [2, 1, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - increaseItemIdx 2`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 3, 
        2],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - increaseItemIdx 3`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - increaseItemIdx 4`] = `
"{
    \\"d\\": null,
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - increaseItemIdx 5`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"d\\": null,
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - increaseItemIdx 6`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"I\\": \\"example\\",
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
  }"
`;

exports[`modifyCodeByCommand - increaseItemIdx 7`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - removeElementFromArray 1`] = `
"{
    \\"a\\": {
      \\"b\\": [2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - removeElementFromArray 2`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - removeElementFromArray 3`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - removeElementFromArray 4`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - removeElementFromArray 5`] = `
"{
    \\"a\\": {
      \\"b\\": [],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - removeElementFromArray 6`] = `
"{
    \\"a\\": {
      \\"b\\": [],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - removeElementFromArray 7`] = `
"{
    \\"a\\": {
      \\"b\\": [],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - removeObjectKey 1`] = `
"{
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - removeObjectKey 2`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - removeObjectKey 3`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }]
  }"
`;

exports[`modifyCodeByCommand - removeObjectKey 4`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - removeObjectKey 5`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4}],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - removeObjectKey 6`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{}],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - removeObjectKey 7`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": true,
    },
    \\"d\\": null,
    \\"e\\": [{}],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - simpleSwap 1`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": false,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - simpleSwap 2`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": \\"hot dog sports\\",
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - simpleSwap 3`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": 0,
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;

exports[`modifyCodeByCommand - simpleSwap 4`] = `
"{
    \\"a\\": {
      \\"b\\": [1, 2, 
        3],
      \\"c\\": {\\"squid\\": [\\"dough\\", false, \\"bag\\", 0]},
    },
    \\"d\\": null,
    \\"e\\": [{ \\"f\\": 4, \\"g\\": 5 }],
    \\"I\\": \\"example\\",
  }"
`;
